python/ray/util/sgd/v2/trainer.py

```diff
@@ -226,8 +234,87 @@ def to_tune_trainable(
         Returns:
             A Trainable that can directly be passed into ``tune.run()``.
         """
+        return _create_tune_trainable(train_func, self._backend,
+                                      self._num_workers, self._use_gpu,
+                                      self._resources_per_worker)
+
 
-        def trainable_func(config: Dict[str, Any]) -> T:
-            pass
+class _TuneTrainer(Trainer):
+    def start_training(
+            self,
+            train_func: Union[Callable[[], T], Callable[[Dict[str, Any]], T]],
+            config: Optional[Dict[str, Any]] = None) -> None:
+        train_func = self._get_train_func(train_func, config)
+        self._executor.start_training(train_func)
+
+    def fetch_next_report(self) -> Optional[List[Dict]]:
+        return self._executor.fetch_next_result()
+
+
+class _SgdTrainable(DistributedTrainable):
+    _function = None
+    _backend = None
+    _num_workers = None
+    _use_gpu = None
+    _resources_per_worker = None
+
+    _started = False
+    _finished = False
+
+    def setup(self, config: Dict):
+        # Create a new Trainer with the same configuration.
+        self._trainer = _TuneTrainer(self._backend, self._num_workers,
+                                     self._use_gpu, self._resources_per_worker)
+        # TODO(matt): use placement group resource allocation
+        self._trainer.start()  # Move this to step?
+        self._config = config
+
+    def step(self) -> Dict:
+        if self._finished:
+            raise RuntimeError("Training has already finished.")
+        if not self._started:
+            self._trainer.start_training(self._function, self._config)
+            self._started = True
+
+        report = self._trainer.fetch_next_report()
+        if report is None:
+            self._finished = True
```

Comment:
Do we have to call `finish_training` here?

Reply:
I updated the PR to add this to be consistent with the regular training flow, but as of now I'm not sure if there are any requirements to calling `finish_training` after exhausting all the reports since we don't process the results. 

One potential benefit I can think of is that it'll help safely clean up the training thread.

Reply:
Ah `finish_training` is needed so that any errors that occur during training will be raised. `fetch_next` doesn't actually raise any errors.

Reply:
Also this would good to add a test for

Reply:
Added a test to make sure a `TuneError` is thrown. Also manually checked that the error is logged in `error.txt` in the trial directory.